### PR TITLE
Refactor runner pipeline

### PR DESCRIPTION
**Pull request by yasassri**
## Purpose
This PR modularize and refactor the shared library code base. This PR fixes parallel execution issue when many builds are running in parallel.

### CODE REVIEW

File: src/org/wso2/tg/jenkins/Constants.groovy [hunk 1]
```diff
@@ -0,0 +1,33 @@
+package org.wso2.tg.jenkins
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/Constants.groovy]: missing license.

File: src/org/wso2/tg/jenkins/PipelineContext.groovy [hunk 2]
```diff
@@ -15,8 +15,19 @@
...
+    static def context
+
+    static def setContext(def context) {
+        this.context = context
+    }
+
+    static def getContext() {
+        return this.context
+    }
+}
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/PipelineContext.groovy]: missing new line EOF.

File: src/org/wso2/tg/jenkins/Properties.groovy [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.wso2.tg.jenkins
+
+@Singleton
+class Properties {
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/Properties.groovy]: -1 to add two classes with seemingly similar functionality (Constants vs Properties).

File: src/org/wso2/tg/jenkins/Properties.groovy [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+    static def TESTGRID_NAME                = "WSO2-TestGrid"
+    static def TESTGRID_DIST_LOCATION       = "/testgrid/testgrid-home/testgrid-dist/"
+    static def TESTGRID_HOME                = "/testgrid/testgrid-home/"
+    static def JOB_CONFIG_YAML              = "job-config.yaml"
+    static def SQL_DRIVERS_LOCATION_UNIX    ="/home/centos/sql-drivers/"
+    static def SQL_DRIVERS_LOCATION_WINDOWS ="/testgrid/sql-drivers"
+    static def REMOTE_WORKSPACE_DIR_UNIX    ="/opt/wso2/workspace"
+    static def REMOTE_WORKSPACE_DIR_WINDOWS ="c:/testgrid/workspace"
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/Properties.groovy]: Groovy can handle the path separator differences?

File: src/org/wso2/tg/jenkins/Properties.groovy [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    private def getCredentials(def key, boolean isMandatory = true){
+        def ctx = PipelineContext.getContext()
+        def cred = ctx.credentials(key).toString()
+        if (cred == null || cred.trim() == "" && isMandatory) {
+            ctx.echo "A mandatory credential is empty or null " + key
+            throw new Exception("A mandatory property " + key + " is empty or null")
+        }
+        ctx.echo "Credential for key : " + key + " is found."
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/Properties.groovy]: Seems this log is too verbose to be a INFO log. Make it debug log instead?

File: src/org/wso2/tg/jenkins/executors/TestExecutor.groovy [hunk 6]
```diff
@@ -18,108 +18,88 @@
...
-        git clone ${SCENARIOS_REPOSITORY}
+    def fileUtil = new FileUtils()
+    def props = Properties.instance
+    def tgExecutor = new TestGridExecutor()
+    def runtime =  new RuntimeUtils()
 
-        echo Cloning ${INFRASTRUCTURE_REPOSITORY} into ${PWD}/${testPlanId}/${INFRA_LOCATION}
-        git clone ${INFRASTRUCTURE_REPOSITORY}
+    fileUtil.createDirectory("${props.WORKSPACE}/${testPlanId}")
+    echo "Preparing workspace"
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/executors/TestExecutor.groovy]: Add more context such as testplan id.

File: src/org/wso2/tg/jenkins/executors/TestExecutor.groovy [hunk 7]
```diff
@@ -18,108 +18,88 @@
...
                                 int fileNo = i
-                                testplanId = commonUtils.getTestPlanId("${PWD}/test-plans/" + files[fileNo].name)
+                                testplanId = commonUtils.getTestPlanId("${props.WORKSPACE}/test-plans/" + files[fileNo].name)
                                 runPlan(files[i], testplanId)
                             }
                         } else {
                             for (int i = 0; i < processFileCount; i++) {
                                 int fileNo = processFileCount * (executor - 1) + i
-                                testplanId = commonUtils.getTestPlanId("${PWD}/test-plans/" + files[fileNo].name)
+                                testplanId = commonUtils.getTestPlanId("${props.WORKSPACE}/test-plans/" + files[fileNo].name)
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/executors/TestExecutor.groovy]: 120 should be a line length in groovy code as well.

File: src/org/wso2/tg/jenkins/util/AWSUtils.groovy [hunk 8]
```diff
@@ -18,31 +18,33 @@
...
 package org.wso2.tg.jenkins.util
 
+import org.wso2.tg.jenkins.Properties
+
 def uploadToS3(testPlanId) {
-    def s3BucketName = getS3BucketName()
+    def props = Properties.instance
     sh """
-      aws s3 sync ${TESTGRID_HOME}/jobs/${PRODUCT}/${testPlanId}/ s3://${s3BucketName}/artifacts/jobs/${PRODUCT}/builds/${testPlanId} --include "*" --exclude 'workspace/*'
+      aws s3 sync ${props.TESTGRID_HOME}/jobs/${props.PRODUCT}/${testPlanId}/ ${getS3WorkspaceURL()}/artifacts/jobs/${props.PRODUCT}/builds/${testPlanId} --include "*" --exclude 'workspace/*'
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/util/AWSUtils.groovy]: 120 line length

File: src/org/wso2/tg/jenkins/util/AWSUtils.groovy [hunk 9]
```diff
@@ -18,31 +18,33 @@
...
+private def getS3BucketName() {
+    def props = Properties.instance
+    def properties = readProperties file: "${props.CONFIG_PROPERTY_FILE_PATH}"
     def bucket = properties['AWS_S3_BUCKET_NAME']
     if ("${bucket}" == "null") {
         bucket = "unknown"
     }
     return bucket
-}
+}
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/util/AWSUtils.groovy]: missing new line eof.

File: src/org/wso2/tg/jenkins/util/EnvironmentUtils.groovy [hunk 10]
```diff
@@ -0,0 +1,19 @@
+package org.wso2.tg.jenkins.util
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/util/EnvironmentUtils.groovy]: missing license.

File: src/org/wso2/tg/jenkins/util/EnvironmentUtils.groovy [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+import hudson.AbortException
+import org.wso2.tg.jenkins.PipelineContext
+
+
+class EnvironmentUtils implements Serializable {
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/util/EnvironmentUtils.groovy]: missing javadocs (or is it called groovydocs? :) ).

File: src/org/wso2/tg/jenkins/util/WorkSpaceUtils.groovy [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+ *
+ * @param filePath full qualified path Of directory
+ * @throws IOException if file creation fails
+ */
+def createJobConfigYamlFile(filePath) throws IOException {
+
+    def props = Properties.instance
+    // https://jenkins.io/doc/pipeline/steps/pipeline-utility-steps/#writeyaml-write-a-yaml-from-an-object
+//    File file = new File (filePath)
+//    file.createNewFile()
```
Comment (kasunbg) [inline src/org/wso2/tg/jenkins/util/WorkSpaceUtils.groovy]: code comments.

File: vars/Runner.groovy [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+                    }
+                }
+            }
+
+            stage('parallel-run') {
+                steps {
+                    script {
+                        def name = "unknown"
+                        try {
+                            parallel_executor_count = 12
```
Comment (kasunbg) [inline vars/Runner.groovy]: hard-coded value!

File: src/org/wso2/tg/jenkins/Properties.groovy (hunk 3 above)
Comment (yasassri) [reply src/org/wso2/tg/jenkins/Properties.groovy]: Removed the Constants class.

File: src/org/wso2/tg/jenkins/Properties.groovy (hunk 4 above)
Comment (yasassri) [reply src/org/wso2/tg/jenkins/Properties.groovy]: This is not used with in groovy rather it is used in the python scripts.

File: src/org/wso2/tg/jenkins/executors/TestExecutor.groovy (hunk 6 above)
Comment (yasassri) [reply src/org/wso2/tg/jenkins/executors/TestExecutor.groovy]: Added.

**Pull request merged**